### PR TITLE
feat!: adds availability_type for read_replicas

### PR DESCRIPTION
**Pull request by g-awmalik**
Addresses issue #328

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: build/int.cloudbuild.yaml [hunk 1]
```diff
@@ -36,35 +36,35 @@ steps:
...
-    waitFor:
-      - apply mssql-ha-local
-    name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
-    args: ['/bin/bash', '-c', 'cft test run TestMsSqlHaModule --stage verify --verbose']
-  - id: teardown mssql-ha-local
-    waitFor:
-      - verify mssql-ha-local
-    name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
-    args: ['/bin/bash', '-c', 'cft test run TestMsSqlHaModule --stage teardown --verbose']
+  # - id: apply mssql-ha-local
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply build/int.cloudbuild.yaml:39]: Yeah I know, seems like there's an upstream [issue](https://github.com/hashicorp/terraform-provider-google/issues/11891) with the provider. I have it commented out to rule that out.

File: modules/mysql/read_replica.tf [hunk 2]
```diff
@@ -38,6 +38,7 @@ resource "google_sql_database_instance" "replicas" {
   settings {
     tier              = lookup(each.value, "tier", var.tier)
     activation_policy = "ALWAYS"
+    availability_type = var.availability_type
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply modules/mysql/read_replica.tf:41]: Yes it should. I'll do that instead.

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/mysql/README.md /tmp/tmp.ttz2sYdgGp/generate_docs/workspace/modules/mysql/README.md
41c41
< | read\_replicas | List of read replicas to create. Encryption key is required for replica in different region. For replica in same region as master set encryption\_key\_name = null | <pre>list(object({<br>    name                  = string<br>    tier                  = string<br>    zone                  = string<br>    disk_type             = string<br>    disk_autoresize       = bool<br>    disk_autoresize_limit = number<br>    disk_size             = string<br>    user_labels           = map(string)<br>    database_flags = list(object({<br>      name  = string<br>      value = string<br>    }))<br>    ip_configuration = object({<br>      authorized_networks = list(map(string))<br>      ipv4_enabled        = bool<br>      private_network     = string<br>      require_ssl         = bool<br>      allocated_ip_range  = string<br>    })<br>    encryption_key_name = string<br>  }))</pre> | `[]` | no |
---
> | read\_replicas | List of read replicas to create. Encryption key is required for replica in different region. For replica in same region as master set encryption\_key\_name = null | <pre>list(object({<br>    name                  = string<br>    tier                  = string<br>    zone                  = string<br>    availability_type     = string<br>    disk_type             = string<br>    disk_autoresize       = bool<br>    disk_autoresize_limit = number<br>    disk_size             = string<br>    user_labels           = map(string)<br>    database_flags = list(object({<br>      name  = string<br>      value = string<br>    }))<br>    ip_configuration = object({<br>      authorized_networks = list(map(string))<br>      ipv4_enabled        = bool<br>      private_network     = string<br>      require_ssl         = bool<br>      allocated_ip_range  = string<br>    })<br>    encryption_key_name = string<br>  }))</pre> | `[]` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/postgresql/README.md /tmp/tmp.ttz2sYdgGp/generate_docs/workspace/modules/postgresql/README.md
43c43
< | read\_replicas | List of read replicas to create. Encryption key is required for replica in different region. For replica in same region as master set encryption\_key\_name = null | <pre>list(object({<br>    name                  = string<br>    tier                  = string<br>    zone                  = string<br>    disk_type             = string<br>    disk_autoresize       = bool<br>    disk_autoresize_limit = number<br>    disk_size             = string<br>    user_labels           = map(string)<br>    database_flags = list(object({<br>      name  = string<br>      value = string<br>    }))<br>    ip_configuration = object({<br>      authorized_networks = list(map(string))<br>      ipv4_enabled        = bool<br>      private_network     = string<br>      require_ssl         = bool<br>      allocated_ip_range  = string<br>    })<br>    encryption_key_name = string<br>  }))</pre> | `[]` | no |
---
> | read\_replicas | List of read replicas to create. Encryption key is required for replica in different region. For replica in same region as master set encryption\_key\_name = null | <pre>list(object({<br>    name                  = string<br>    tier                  = string<br>    availability_type     = string<br>    zone                  = string<br>    disk_type             = string<br>    disk_autoresize       = bool<br>    disk_autoresize_limit = number<br>    disk_size             = string<br>    user_labels           = map(string)<br>    database_flags = list(object({<br>      name  = string<br>      value = string<br>    }))<br>    ip_configuration = object({<br>      authorized_networks = list(map(string))<br>      ipv4_enabled        = bool<br>      private_network     = string<br>      require_ssl         = bool<br>      allocated_ip_range  = string<br>    })<br>    encryption_key_name = string<br>  }))</pre> | `[]` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate ./examples/mssql-public
Success!
 The configuration is valid.
terraform_validate ./examples/mysql-backup-create-service-account
Success!
 The configuration is valid.
terraform_validate ./examples/mysql-ha
Success!
 The configuration is valid.
terraform_validate ./examples/mysql-private
╷
│
 
Error: 
Invalid value for module argument
│
 
│
 
  on ../../modules/safer_mysql/main.tf line 71, in module "safer_mysql":
│
 
  71:   read_replicas                    = 
var.read_replicas
│
 
│
 
The given value is not suitable for child module variable "read_replicas"
│
 
defined at ../../modules/mysql/variables.tf:175,1-25: incorrect list
│
 
element type: attribute "availability_type" is required.
╵
terraform_validate ./examples/mysql-public
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-backup-provided-service-account
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-ha
╷
│
 
Error: 
Invalid value for module argument
│
 
│
 
  on main.tf line 82, in module "pg":
│
 
  82:   read_replicas = 
[
│
 
  83:     {
│
 
  84:       name                  = "0"
│
 
  85:       zone                  = "us-central1-a"
│
 
  86:       tier                  = "db-custom-1-3840"
│
 
  87:       ip_configuration      = local.read_replica_ip_configuration
│
 
  88:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
  89:       disk_autoresize       = null
│
 
  90:       disk_autoresize_limit = null
│
 
  91:       disk_size             = null
│
 
  92:       disk_type             = "PD_HDD"
│
 
  93:       user_labels           = { bar = "baz" }
│
 
  94:       encryption_key_name   = null
│
 
  95:     },
│
 
  96:     {
│
 
  97:       name                  = "1"
│
 
  98:       zone                  = "us-central1-b"
│
 
  99:       tier                  = "db-custom-1-3840"
│
 
 100:       ip_configuration      = local.read_replica_ip_configuration
│
 
 101:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
 102:       disk_autoresize       = null
│
 
 103:       disk_autoresize_limit = null
│
 
 104:       disk_size             = null
│
 
 105:       disk_type             = "PD_HDD"
│
 
 106:       user_labels           = { bar = "baz" }
│
 
 107:       encryption_key_name   = null
│
 
 108:     },
│
 
 109:     {
│
 
 110:       name                  = "2"
│
 
 111:       zone                  = "us-central1-c"
│
 
 112:       tier                  = "db-custom-1-3840"
│
 
 113:       ip_configuration      = local.read_replica_ip_configuration
│
 
 114:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
 115:       disk_autoresize       = null
│
 
 116:       disk_autoresize_limit = null
│
 
 117:       disk_size             = null
│
 
 118:       disk_type             = "PD_HDD"
│
 
 119:       user_labels           = { bar = "baz" }
│
 
 120:       encryption_key_name   = null
│
 
 121:     },
│
 
 122:   ]
│
 
│
 
The given value is not suitable for child module variable "read_replicas"
│
 
defined at ../../modules/postgresql/variables.tf:182,1-25: element 0:
│
 
attribute "availability_type" is required.
╵
terraform_validate ./examples/postgresql-public
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-public-iam
Success!
 The configuration is valid.
terraform_validate ./modules/backup
Success!
 The configuration is valid.
terraform_validate ./modules/mssql
Success!
 The configuration is valid.
terraform_validate ./modules/mysql
Success!
 The configuration is valid.
terraform_validate ./modules/postgresql
Success!
 The configuration is valid.
terraform_validate ./modules/private_service_access
Success!
 The configuration is valid.
terraform_validate ./modules/safer_mysql
╷
│
 
Error: 
Invalid value for module argument
│
 
│
 
  on main.tf line 71, in module "safer_mysql":
│
 
  71:   read_replicas                    = 
var.read_replicas
│
 
│
 
The given value is not suitable for child module variable "read_replicas"
│
 
defined at ../mysql/variables.tf:175,1-25: incorrect list element type:
│
 
attribute "availability_type" is required.
╵
terraform_validate ./test/fixtures/mssql-ha
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mssql-public
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mysql-ha
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mysql-private
╷
│
 
Error: 
Invalid value for module argument
│
 
│
 
  on ../../../modules/safer_mysql/main.tf line 71, in module "safer_mysql":
│
 
  71:   read_replicas                    = 
var.read_replicas
│
 
│
 
The given value is not suitable for child module variable "read_replicas"
│
 
defined at ../../../modules/mysql/variables.tf:175,1-25: incorrect list
│
 
element type: attribute "availability_type" is required.
╵
terraform_validate ./test/fixtures/mysql-public
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/postgresql-ha
╷
│
 
Error: 
Invalid value for module argument
│
 
│
 
  on ../../../examples/postgresql-ha/main.tf line 82, in module "pg":
│
 
  82:   read_replicas = 
[
│
 
  83:     {
│
 
  84:       name                  = "0"
│
 
  85:       zone                  = "us-central1-a"
│
 
  86:       tier                  = "db-custom-1-3840"
│
 
  87:       ip_configuration      = local.read_replica_ip_configuration
│
 
  88:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
  89:       disk_autoresize       = null
│
 
  90:       disk_autoresize_limit = null
│
 
  91:       disk_size             = null
│
 
  92:       disk_type             = "PD_HDD"
│
 
  93:       user_labels           = { bar = "baz" }
│
 
  94:       encryption_key_name   = null
│
 
  95:     },
│
 
  96:     {
│
 
  97:       name                  = "1"
│
 
  98:       zone                  = "us-central1-b"
│
 
  99:       tier                  = "db-custom-1-3840"
│
 
 100:       ip_configuration      = local.read_replica_ip_configuration
│
 
 101:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
 102:       disk_autoresize       = null
│
 
 103:       disk_autoresize_limit = null
│
 
 104:       disk_size             = null
│
 
 105:       disk_type             = "PD_HDD"
│
 
 106:       user_labels           = { bar = "baz" }
│
 
 107:       encryption_key_name   = null
│
 
 108:     },
│
 
 109:     {
│
 
 110:       name                  = "2"
│
 
 111:       zone                  = "us-central1-c"
│
 
 112:       tier                  = "db-custom-1-3840"
│
 
 113:       ip_configuration      = local.read_replica_ip_configuration
│
 
 114:       database_flags        = [{ name = "autovacuum", value = "off" }]
│
 
 115:       disk_autoresize       = null
│
 
 116:       disk_autoresize_limit = null
│
 
 117:       disk_size             = null
│
 
 118:       disk_type             = "PD_HDD"
│
 
 119:       user_labels           = { bar = "baz" }
│
 
 120:       encryption_key_name   = null
│
 
 121:     },
│
 
 122:   ]
│
 
│
 
The given value is not suitable for child module variable "read_replicas"
│
 
defined at ../../../modules/postgresql/variables.tf:182,1-25: element 0:
│
 
attribute "availability_type" is required.
╵
terraform_validate ./test/fixtures/postgresql-public
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/postgresql-public-iam
Success!
 The configuration is valid.
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by enocom** [comment]
@g-awmalik thanks for this!

File: build/int.cloudbuild.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bharathkkb) [reply build/int.cloudbuild.yaml:39]: Gotcha, in that case lets pin to the LKG provider version in the example and open an issue to track that upstream fix.

File: modules/mysql/variables.tf [hunk 3]
```diff
@@ -178,6 +178,7 @@ variable "read_replicas" {
     name                  = string
     tier                  = string
     zone                  = string
+    availability_type     = string
```
Comment (bharathkkb) [inline modules/mysql/variables.tf:181]: note: breaking

File: examples/postgresql-ha/main.tf [hunk 4]
```diff
@@ -83,6 +83,7 @@ module "pg" {
     {
       name                  = "0"
       zone                  = "us-central1-a"
+      availability_type     = "ZONAL"
```
Comment (bharathkkb) [inline examples/postgresql-ha/main.tf:86]: ZONAL seems to be the default anyway, lets set one to the other value and assert expected in the test.

**Review by bharathkkb — APPROVED** [review]
Minor comments

File: modules/mysql/variables.tf (hunk 3 above)
Comment (bharathkkb) [reply modules/mysql/variables.tf:181]: lets add an upgrade guide: https://github.com/terraform-google-modules/terraform-google-sql-db/tree/master/docs

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: build/int.cloudbuild.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (g-awmalik) [reply build/int.cloudbuild.yaml]: https://github.com/terraform-google-modules/terraform-google-sql-db/issues/330

File: modules/mysql/variables.tf (hunk 3 above)
Comment (g-awmalik) [reply modules/mysql/variables.tf:181]: done

File: examples/postgresql-ha/main.tf (hunk 4 above)
Comment (g-awmalik) [reply examples/postgresql-ha/main.tf]: Set this to REGIONAL and asserted accordingly.

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
test/fixtures/mssql-public/main.tf
--- old/test/fixtures/mssql-public/main.tf
+++ new/test/fixtures/mssql-public/main.tf
@@ -17,7 +17,7 @@
 terraform {
   required_providers {
     google = {
-      source  = "hashicorp/google"
+      source = "hashicorp/google"
       /*
         Pinning this version due to an upstream provider issue: https://github.com/hashicorp/terraform-provider-google/issues/11891
       */
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
test/fixtures/mssql-ha/main.tf
--- old/test/fixtures/mssql-ha/main.tf
+++ new/test/fixtures/mssql-ha/main.tf
@@ -17,7 +17,7 @@
 terraform {
   required_providers {
     google = {
-      source  = "hashicorp/google"
+      source = "hashicorp/google"
       /*
         Pinning this version due to an upstream provider issue: https://github.com/hashicorp/terraform-provider-google/issues/11891
       */
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@g-awmalik
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: modules/postgresql/main.tf [hunk 5]
```diff
@@ -31,6 +31,10 @@ locals {
     is_account_sa = trimsuffix(iu, "gserviceaccount.com") == iu ? false : true
   }]
 
+  // HA method using REGIONAL availability_type requires point in time recovery to be enabled
```
Comment (mans0954) [inline modules/postgresql/main.tf:34]: @g-awmalik Do you have a reference for this?

**Pull request merged**